sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BeamRowToStorageApiProto.java

```diff
@@ -291,9 +307,17 @@ private static Object toProtoValue(
         if (arrayElementType == null) {
           throw new RuntimeException("Unexpected null element type!");
         }
-        return list.stream()
-            .map(v -> toProtoValue(fieldDescriptor, arrayElementType, v))
-            .collect(Collectors.toList());
+        boolean shouldFlatMap =
+            arrayElementType.getTypeName().isCollectionType()
+                || arrayElementType.getTypeName().isMapType();
+
+        Stream<Object> valueStream =
+            list.stream().map(v -> toProtoValue(fieldDescriptor, arrayElementType, v));
+
+        if (shouldFlatMap) {
+          valueStream = valueStream.flatMap(vs -> ((List) vs).stream());
```

Comment:
Why are we introducing this flattening here?

Reply:
See the comment here: https://github.com/apache/beam/pull/22179#discussion_r1037400402

Reply:
Also added a comment on the code to explain it.

Reply:
removed the flattening of nested container types and added a missing check for array<maps>.